### PR TITLE
Servo assembly

### PR DESCRIPTION
**Pull request by RealOrangeOne**
https://trello.com/c/YgyAhMq4/6-write-servo-assembly-documentation

### CODE REVIEW

File: content/api/servo-board/GPIO.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+## Pin Mode
+GPIO pins support 4 different modes, allowing it to work in 4 different ways. A pin can only have 1 mode at a time, and certain pins aren't compatible with certain modes. These Pin modes are stored in an [enum](https://docs.python.org/3/library/enum.html) which needs to be imported to use.
+
+```python
+from robot import PinMode
+``` 
+
+{{% notice note %}}
+The input modes closely resemble those of an Arduino. More information on them can be found in [their docs](https://docs.python.org/3/library/enum.html)
```
Comment (Adimote) [inline content/api/servo-board/GPIO.md]: wrong link?

File: content/api/servo-board/GPIO.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+{{% /notice %}}
+
+
+### `PinMode.INPUT`
+Allows reading the value from the pin
+
+### `PinMode.INPUT_PULLUP`
+Allows reading the value from the pin using the built-in pullup resistor.
```
Comment (Adimote) [inline content/api/servo-board/GPIO.md]: Feels like we need to explain pullup resistors somewhere, or at least link to a good explanation

File: content/api/servo-board/GPIO.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+```
+
+## Reading a pin value
+
+To read the value of a pin, simply call `read()`. This will return either `PinValue.HIGH` or `PinValue.LOW`, which have to be imported to use.
+
+```python
+from robot import PinMode
+
+r.servo_board.gpio[3].read() == PinValue.HIGH
```
Comment (Adimote) [inline content/api/servo-board/GPIO.md]: Need to import PinValue

File: content/api/servo-board/GPIO.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+>>> True
+```
+
+{{% notice warning %}}
+Calling `read()` will raise an exception if the pins mode hasn't been set to an input type beforehand.
+{{% /notice %}}
+
+# Analogue Pins
+
+Certain sensors output analogue signals rather than digital ones, so have to be read differently. The servo assembly has 4 analogue inputs.
```
Comment (Adimote) [inline content/api/servo-board/GPIO.md]: link to / explain analogue signals?

File: content/api/servo-board/_index.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+### Get a specific board
+```python
+board_one = r.servo_boards['S3rv0']
+board_one = r.servo_boards[0]
+```
+
+{{% notice tip %}}
+If you already have a board instance, and want to get its serial number, you can use the `serial` property.
```
Comment (Adimote) [inline content/api/servo-board/_index.md]: avoid words like 'instance', I'd suggest rewording this to "if you want to get the serial for a board you can use the `serial` property, i.e. `print(r.servo_boards[0].serial)`

File: content/api/servo-board/GPIO.md (hunk 2 above)
Comment (RealOrangeOne) [reply content/api/servo-board/GPIO.md]: Agreed, although i'm not entirely sure what one is myself, or at least why it's preferable in some cases. Anyone know a good resource / an provide a short explanation? cc @kierdavis

Comment (kierdavis) [reply content/api/servo-board/GPIO.md]: https://learn.sparkfun.com/tutorials/pull-up-resistors looks like a good resource, and it also talks about internal pullup resistors inside microcontrollers (like we use here).

A disadvantage is that it contains code example(s) for Arduino, which may be confusing since it's a different syntax to our kit. I don't think it's too much of a problem though - especially if we ensure there's an example using our syntax near to this link.

It's not an easy topic to explain succinctly - I feel like any description I could write would still be a few paragraphs long.

File: content/api/servo-board/GPIO.md (hunk 4 above)
Comment (kierdavis) [reply content/api/servo-board/GPIO.md]: Something brief along the lines of

> Analogue signals can have any voltage, while digital signals can only take on one of two voltages. They need to be connected to one of the servo's assembly's 4 dedicated analogue inputs.

would probably do here.

File: content/api/servo-board/GPIO.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
+---
+title: GPIO
+---
+
+You can use the GPIO pins to read and write individual pins of the servo assembly. These can be used for anything, from microswitches to LEDs. GPIO is only available on pins 2 to 12, because pins 0 and 1 are reserved for communication with our kit.
```
Comment (kierdavis) [inline content/api/servo-board/GPIO.md]: I think it'd be a good idea to state what GPIO stands for here.

File: content/api/servo-board/GPIO.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+To read the value of an analogue pin, call `read_analogue()` on the servo board. This will give you the value of all analogue pins.
+
+```python
+r.servo_board.read_analogue()["a0"]
+>>> 1.569
+``` 
+
+{{% notice info %}}
+The values are the voltage read on the pins, between 0 and 5.
```
Comment (kierdavis) [inline content/api/servo-board/GPIO.md]: "voltage" -> "voltages"

File: content/api/servo-board/servos.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+---
+title: Servos
+---
+
+From the servo assembly, you can control the attached servos. By default, servos will be unpowered when your robot starts, and can freely rotate. Once you set a value, they then fix to that rotation. This also happens when you turn off your robot, or unplug your USB.
```
Comment (kierdavis) [inline content/api/servo-board/servos.md]: I'd recommend saying "USB stick" instead of just "USB" (in case it gets misinterpreted as the USB connection to the servo assembly).

File: content/api/servo-board/servos.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+Servos can be controlled using the `position` parameter. This should be set to a value between -1 and 1.  
+
+```python
+servo_one.position = 0.65
+print(servo_one.position)
+>>> 0.65
+```
+
+{{% notice tip %}}
+When `position` is set to 0, the servo loses power and can freely rotate. The same state they're in at start-up.
```
Comment (kierdavis) [inline content/api/servo-board/servos.md]: I believe this occurs when `position` is set to `None`, not `0`.

See: https://github.com/sourcebots/robotd/blob/master/robotd/devices.py#L349

File: content/api/servo-board/ultrasound.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+## Reading a distance
+Assuming you wired the _trigger_ pin to pin 6, and the _echo_ pin to pin 7, you can query the distance from the sensor using:
+
+```python
+r.servo_board.read_ultrasound(6, 7)
+>>> 0.524
+```
+
+{{% notice info %}}
+The returned distance is in meters.
```
Comment (kierdavis) [inline content/api/servo-board/ultrasound.md]: meters -> metres

**Comment by kierdavis** [comment]
Will wait for @Adimote to take a look before merging.

**Comment by kierdavis** [comment]
What's the reasoning behind that last commit? (just curious)

**Comment by RealOrangeOne** [comment]
@kierdavis related to https://github.com/sourcebots/docs/pull/30

**Comment by kierdavis** [comment]
Ahh I see, cool.

Looks like there's a merge conflict with the `.spelling` file now, could you check that out?
